File: typespec/hub/workhistory.ts
export interface AddWorkHistoryRequest {
  employer_domain: string;
  title: string;
  start_date: string;
  end_date?: string;
  description?: string;
}

export interface AddWorkHistoryResponse {
  work_history_id: string;
}

export interface WorkHistory {
  id: string;
  employer_domain: string;
  employer_name?: string;
  employer_logo_url?: string;
  title: string;
  start_date: string;
  end_date?: string;
  description?: string;
}

export interface UpdateWorkHistoryRequest {
  id: string;
  title: string;
  start_date: string;
  end_date?: string;
  description?: string;
}

export interface ListWorkHistoryRequest {
  user_handle?: string;
}

export interface DeleteWorkHistoryRequest {
  id: string;
}

function isValidDate(dateStr: string): boolean {
  const date = new Date(dateStr);
  return date instanceof Date && !isNaN(date.getTime());
}

export function isValidAddWorkHistoryRequest(req: AddWorkHistoryRequest): {
  valid: boolean;
  error?: string;
} {
  if (!req.employer_domain) {
    return { valid: false, error: "employer_domain is required" };
  }
  if (!req.title) {
    return { valid: false, error: "title is required" };
  }
  if (!req.start_date || !isValidDate(req.start_date)) {
    return {
      valid: false,
      error: "start_date is required and must be a valid date",
    };
  }
  if (req.end_date && !isValidDate(req.end_date)) {
    return { valid: false, error: "end_date must be a valid date if provided" };
  }
  if (req.description && req.description.length > 1024) {
    return {
      valid: false,
      error: "description must not exceed 1024 characters",
    };
  }
  return { valid: true };
}

export function isValidUpdateWorkHistoryRequest(
  req: UpdateWorkHistoryRequest
): { valid: boolean; error?: string } {
  if (!req.id) {
    return { valid: false, error: "id is required" };
  }
  if (!req.title) {
    return { valid: false, error: "title is required" };
  }
  if (!req.start_date || !isValidDate(req.start_date)) {
    return {
      valid: false,
      error: "start_date is required and must be a valid date",
    };
  }
  if (req.end_date && !isValidDate(req.end_date)) {
    return { valid: false, error: "end_date must be a valid date if provided" };
  }
  if (req.description && req.description.length > 1024) {
    return {
      valid: false,
      error: "description must not exceed 1024 characters",
    };
  }
  return { valid: true };
}

export function isValidListWorkHistoryRequest(req: ListWorkHistoryRequest): {
  valid: boolean;
  error?: string;
} {
  // No validation needed for user_handle as it's optional and any string is valid
  return { valid: true };
}

File: typespec/hub/workhistory.go
package hub

type AddWorkHistoryRequest struct {
	EmployerDomain string  `json:"employer_domain" validate:"required,validate_domain"`
	Title          string  `json:"title"           validate:"required,min=3,max=64"`
	StartDate      string  `json:"start_date"      validate:"required,datetime=2006-01-02"`
	EndDate        *string `json:"end_date"        validate:"omitempty,datetime=2006-01-02,date_after=StartDate"`
	Description    *string `json:"description"     validate:"omitempty,max=1024"`
}

type AddWorkHistoryResponse struct {
	WorkHistoryID string `json:"work_history_id"`
}

type WorkHistory struct {
	ID             string `json:"id"`
	EmployerDomain string `json:"employer_domain" validate:"required,validate_domain"`

	EmployerName *string `json:"employer_name"`
	EmployerLogo *string `json:"employer_logo"`

	Title       string  `json:"title"       validate:"required,min=3,max=64"`
	StartDate   string  `json:"start_date"  validate:"required,datetime=2006-01-02"`
	EndDate     *string `json:"end_date"    validate:"omitempty,datetime=2006-01-02,date_after=StartDate"`
	Description *string `json:"description" validate:"omitempty,max=1024"`
}

type UpdateWorkHistoryRequest struct {
	ID          string  `json:"id"          validate:"required,uuid"`
	Title       string  `json:"title"       validate:"required,min=3,max=64"`
	StartDate   string  `json:"start_date"  validate:"required,datetime=2006-01-02"`
	EndDate     *string `json:"end_date"    validate:"omitempty,datetime=2006-01-02,date_after=StartDate"`
	Description *string `json:"description" validate:"omitempty,max=1024"`
}

type ListWorkHistoryRequest struct {
	UserHandle *string `json:"user_handle" validate:"omitempty,min=3,max=64"`
}

type DeleteWorkHistoryRequest struct {
	ID string `json:"id" validate:"required,uuid"`
}
